File: Infer.test.ts
import { assertEquals } from "https://deno.land/std@0.137.0/testing/asserts.ts";
import { inferExpression } from "./Infer.ts";
import { parse } from "./Parser.ts";
import {
  emptyTypeEnv,
  Scheme,
  TArr,
  TCon,
  TVar,
  Type,
  typeBool,
  typeInt,
} from "./Typing.ts";

Deno.test("infer Apply", () => {
  const [constraints, type] = inferExpression(
    emptyTypeEnv
      .extend("a", new Scheme(["T"], new TArr(new TVar("T"), new TVar("T"))))
      .extend("b", new Scheme([], typeInt)),
    parse("a b").expr,
  );

  assertEquals(constraints.constraints.length, 1);
  assertEquals(constraints.constraints[0], [
    new TArr(new TVar("V1"), new TVar("V1")),
    new TArr(new TCon("Int"), new TVar("V2")),
  ]);
  assertEquals(type, new TVar("V2"));
});

Deno.test("infer If", () => {
  const [constraints, type] = inferExpression(
    emptyTypeEnv
      .extend("a", new Scheme(["S"], new TVar("S")))
      .extend("b", new Scheme([], typeInt))
      .extend("c", new Scheme(["T"], new TVar("T"))),
    parse("if (a) b else c").expr,
  );

  assertEquals(constraints.constraints.length, 2);
  assertEquals(constraints.constraints[0], [new TVar("V1"), typeBool]);
  assertEquals(constraints.constraints[1], [typeInt, new TVar("V2")]);
  assertEquals(type, typeInt);
});

Deno.test("infer LBool", () => {
  const [constraints, type] = inferExpression(
    emptyTypeEnv,
    parse("True").expr,
  );

  assertEquals(constraints.constraints.length, 0);
  assertEquals(type, typeBool);
});

Deno.test("infer Lam", () => {
  const [constraints, type] = inferExpression(
    emptyTypeEnv,
    parse("\\x -> x 10").expr,
  );

  assertEquals(constraints.constraints.length, 1);
  assertEquals(constraints.constraints[0], [
    new TVar("V1"),
    new TArr(typeInt, new TVar("V2")),
  ]);
  assertEquals(type, new TArr(new TVar("V1"), new TVar("V2")));
});

Deno.test("infer Let", () => {
  const [constraints, type] = inferExpression(
    emptyTypeEnv,
    parse("let x = 10 in x").expr,
  );

  assertEquals(constraints.constraints.length, 0);
  assertEquals(type, typeInt);
});

Deno.test("infer LInt", () => {
  const [constraints, type] = inferExpression(
    emptyTypeEnv,
    parse("123").expr,
  );

  assertEquals(constraints.constraints.length, 0);
  assertEquals(type, typeInt);
});

Deno.test("infer Op", () => {
  const scenario = (input: string, resultType: Type) => {
    const [constraints, type] = inferExpression(
      emptyTypeEnv
        .extend("a", new Scheme(["T"], new TVar("T")))
        .extend("b", new Scheme(["T"], new TVar("T"))),
      parse(input).expr,
    );
    assertEquals(constraints.constraints.length, 1);

    assertEquals(
      constraints.constraints[0],
      [
        new TArr(new TVar("V1"), new TArr(new TVar("V2"), new TVar("V3"))),
        new TArr(typeInt, new TArr(typeInt, resultType)),
      ],
    );
    assertEquals(type, new TVar("V3"));
  };

  scenario("a + b", typeInt);
  scenario("a - b", typeInt);
  scenario("a * b", typeInt);
  scenario("a / b", typeInt);
  scenario("a == b", typeBool);
});

Deno.test("infer Var", () => {
  const [constraints, type] = inferExpression(
    emptyTypeEnv.extend(
      "a",
      new Scheme(["T"], new TArr(new TVar("T"), new TVar("T"))),
    ),
    parse("a").expr,
  );

  assertEquals(constraints.constraints.length, 0);
  assertEquals(type, new TArr(new TVar("V1"), new TVar("V1")));
});

File: kotlin-tlca/app/src/test/kotlin/tlca/InferTest.kt
package tlca

import kotlin.test.Test
import kotlin.test.assertEquals

class InferTest {
    @Test
    fun inferApply() {
        assertInference(
            emptyTypeEnv
                    + Pair("a", Scheme(setOf("T"), TArr(TVar("T"), TVar("T"))))
                    + Pair("b", Scheme(emptySet(), typeInt)),
            "a b",
            listOf("V1 -> V1 ~ Int -> V2"),
            listOf("V2")
        )
    }

    @Test
    fun inferIf() {
        assertInference(
            emptyTypeEnv
                    + Pair("a", Scheme(setOf("S"), TVar("S")))
                    + Pair("b", Scheme(emptySet(), typeInt))
                    + Pair("c", Scheme(setOf("S"), TVar("S"))),
            "if (a) b else c",
            listOf(
                "V1 ~ Bool",
                "Int ~ V2"
            ),
            listOf("Int")
        )
    }

    @Test
    fun inferLam() {
        assertInference(
            emptyTypeEnv,
            "\\x -> x 10",
            listOf("V1 ~ Int -> V2"),
            listOf("V1 -> V2")
        )
    }

    @Test
    fun inferLBool() {
        assertInference(emptyTypeEnv, "True", emptyList(), listOf("Bool"))
    }

    @Test
    fun inferLInt() {
        assertInference(emptyTypeEnv, "123", emptyList(), listOf("Int"))
    }

    @Test
    fun inferLet() {
        assertInference(
            emptyTypeEnv,
            "let x = 10 and y = x + 1 ; y",
            emptyList(),
            listOf("(Int * Int)", "Int")
        )
    }

    @Test
    fun inferOp() {
        fun scenario(input: String, resultType: Type) {
            assertInference(
                emptyTypeEnv
                        + Pair("a", Scheme(setOf("T"), TVar("T")))
                        + Pair("b", Scheme(setOf("T"), TVar("T"))),
                input,
                listOf(
                    "V1 -> V2 -> V3 ~ Int -> Int -> $resultType"
                ),
                listOf("V3")
            )
        }

        scenario("a + b", typeInt)
        scenario("a - b", typeInt)
        scenario("a * b", typeInt)
        scenario("a / b", typeInt)
        scenario("a == b", typeBool)
    }

    @Test
    fun inferVar() {
        assertInference(
            emptyTypeEnv + Pair("a", Scheme(setOf("T"), TArr(TVar("T"), TVar("T")))),
            "a",
            emptyList(),
            listOf("V1 -> V1")
        )
    }

    private fun assertInference(typeEnv: TypeEnv, input: String, expectedConstraints: List<String>, expectedTypes: List<String>) {
        val inferResult = infer(typeEnv, parse(input), Constraints(), Pump())

        assertConstraints(inferResult.constraints, expectedConstraints)
        assertEquals(expectedTypes, inferResult.types.map { it.toString() })
    }

    private fun assertConstraints(constraints: Constraints, expected: List<String>) {
        assertEquals(expected.joinToString(", "), constraints.toString())
    }
}
